python/lsst/pipe/base/quantum_graph/aggregator/_communicators.py

```diff
@@ -0,0 +1,970 @@
+        """Make an empty queue that can be used to pass objects between
+        workers in this context.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def make_event(self) -> Event:
+        """Make an empty that can be used to communicate a boolean state change
+        to workers in this context.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def make_worker(
+        self, target: Callable[..., None], args: tuple[Any, ...], name: str | None = None
+    ) -> Worker:
+        """Make a worker that runs the given callable.
+
+        Parameters
+        ----------
+        target : `~collections.abc.Callable`
+            A callable to invoke on the worker.
+        args : `tuple`
+            Positional arguments to pass to the callable.
+        name : `str`, optional
+            Human-readable name for the worker.
+
+        Returns
+        -------
+        worker : `threading.Thread` or `multiprocessing.Process`
+            Process or thread.  Will need to have its ``start`` method called
+            to actually begin.
+        """
+        raise NotImplementedError()
+
+
+class ThreadingContext(WorkerContext):
+    """An implementation of `WorkerContext` backed by the `threading`
+    module.
+    """
+
+    def make_queue(self) -> Queue[Any]:
+        return queue.Queue()
+
+    def make_event(self) -> Event:
+        return threading.Event()
+
+    def make_worker(
+        self, target: Callable[..., None], args: tuple[Any, ...], name: str | None = None
+    ) -> Worker:
+        return threading.Thread(target=target, args=args, name=name)
+
+
+class SpawnProcessContext(WorkerContext):
+    """An implementation of `WorkerContext` backed by the `multiprocessing`
+    module, with new processes started by spawning.
+    """
+
+    def __init__(self) -> None:
+        self._ctx = multiprocessing.get_context("spawn")
+
+    def make_queue(self) -> Queue[Any]:
+        return self._ctx.Queue()
+
+    def make_event(self) -> Event:
+        return self._ctx.Event()
+
+    def make_worker(
+        self, target: Callable[..., None], args: tuple[Any, ...], name: str | None = None
+    ) -> Worker:
+        return self._ctx.Process(target=target, args=args, name=name)
+
+
+def _get_from_queue(q: Queue[_T], block: bool = False, timeout: float | None = None) -> _T | None:
+    """Get an object from a queue and return `None` if it is empty.
+
+    Parameters
+    ----------
+    q : `Queue`
+        Queue to get an object from.
+    block : `bool`
+        Whether to block until an object is available.
+    timeout : `float` or `None`, optional
+        Maximum number of seconds to wait while blocking.
+
+    Returns
+    -------
+    obj : `object` or `None`
+        Object from the queue, or `None` if it was empty.
+    """
+    try:
+        return q.get(block=block, timeout=timeout)
+    except queue.Empty:
+        return None
+
+
+class CancelError(BaseException):
```

Comment:
I read this as an error occurred during a cancel operation.  Perhaps something like `CriticalWorkerError` or `FatalApplicationError`